File: engines/web/src/resource-helpers.ts
// Copyright (c) 2026 Dennis Brandl
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
import type { ResourceCommandSpecification } from './types.js';

const ACTIVATION_COMMANDS = new Set([
  'Acquire', 'Acquire Pool Amount', 'Send', 'Receive', 'Synchronize',
]);

const SYNC_COMMANDS = new Set(['Send', 'Receive', 'Synchronize']);

export function splitResourceCommands(
  commands: ResourceCommandSpecification[],
): { activation: ResourceCommandSpecification[]; completion: ResourceCommandSpecification[] } {
  const activation: ResourceCommandSpecification[] = [];
  const completion: ResourceCommandSpecification[] = [];
  for (const cmd of commands) {
    if (ACTIVATION_COMMANDS.has(cmd.command_type)) {
      activation.push(cmd);
    } else {
      completion.push(cmd);
    }
  }
  return { activation, completion };
}

export function sortActivationCommands(
  commands: ResourceCommandSpecification[],
): ResourceCommandSpecification[] {
  return [...commands].sort((a, b) => {
    const aIsSync = SYNC_COMMANDS.has(a.command_type) ? 0 : 1;
    const bIsSync = SYNC_COMMANDS.has(b.command_type) ? 0 : 1;
    if (aIsSync !== bIsSync) return aIsSync - bIsSync;
    return a.resource_name.localeCompare(b.resource_name);
  });
}

File: engines/kmp-engine/src/commonMain/kotlin/com/trajectoryruntime/engine/ResourceHelpers.kt
// Copyright (c) 2026 Dennis Brandl
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
package com.trajectoryruntime.engine

private val ACTIVATION_COMMANDS = setOf(
    "Acquire", "Acquire Pool Amount", "Send", "Receive", "Synchronize",
)

private val SYNC_COMMANDS = setOf("Send", "Receive", "Synchronize")

data class SplitCommands(
    val activation: List<ResourceCommandSpecification>,
    val completion: List<ResourceCommandSpecification>,
)

fun splitResourceCommands(commands: List<ResourceCommandSpecification>): SplitCommands {
    val activation = mutableListOf<ResourceCommandSpecification>()
    val completion = mutableListOf<ResourceCommandSpecification>()
    for (cmd in commands) {
        if (cmd.command_type in ACTIVATION_COMMANDS) {
            activation.add(cmd)
        } else {
            completion.add(cmd)
        }
    }
    return SplitCommands(activation, completion)
}

fun sortActivationCommands(
    commands: List<ResourceCommandSpecification>,
): List<ResourceCommandSpecification> {
    return commands.sortedWith(compareBy(
        { if (it.command_type in SYNC_COMMANDS) 0 else 1 },
        { it.resource_name },
    ))
}
